internal/tasks/device_render.go

```diff
@@ -176,6 +185,28 @@ func renderConfigItems(ctx context.Context, config *[]api.DeviceSpec_Config_Item
 	return nil
 }
 
+func validateConfigParameters(configItem *api.DeviceSpec_Config_Item, args *renderConfigArgs) error {
+	cfgJson, err := configItem.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("failed converting configuration to json: %w", err)
+	}
+	if args.validateOnly {
+		// Make sure all parameters are in the proper format
+		return ValidateParameterFormat(cfgJson)
+	}
+
+	// If we're rendering the device config and it still has parameters, something went wrong
+	if ContainsParameter(cfgJson) {
+		if args.deviceBelongsToFleet {
+			return fmt.Errorf("configuration contains parameter, perhaps due to a missing device label")
```

Comment:
I've been trying to get the system into this error, as it appears like it could fix the bug I created earlier (EDM-303).

However, in none of the combinations this has happened (by creating the device before the fleet, or viceversa, or by editing the fleet config after the device and fleet are bound), and I believe the scenario is the one you're trying to cover here.

Fleet
```
{
  "apiVersion": "v1alpha1",
  "kind": "Fleet",
  "metadata": {
    "annotations": {
      "fleet-controller/templateVersion": "2024-08-06T12:34:01Z"
    },
    "creationTimestamp": "2024-08-06T12:33:18.869517Z",
    "generation": 2,
    "labels": {},
    "name": "empty-fleet",
    "resourceVersion": "6"
  },
  "spec": {
    "selector": {
      "matchLabels": {
        "device": "device-a"
      }
    },
    "template": {
      "metadata": {
        "generation": 2,
        "labels": {
          "fleet": "empty-fleet"
        }
      },
      "spec": {
        "config": [
          {
            "configType": "GitConfigProviderSpec",
            "gitRef": {
              "path": "/demos/{{device.metadata.labels[demo]}}/configuration/etc/microshift/manifests",
              "repository": "default-repo",
              "targetRevision": "main"
            },
            "name": "test-add-after-create"
          }
        ]
      }
    }
  },
  "status": {
    "conditions": [
      {
        "lastTransitionTime": "2024-08-06T12:33:18.88760862Z",
        "message": "",
        "reason": "Valid",
        "status": "True",
        "type": "Valid"
      },
      {
        "lastTransitionTime": "2024-08-06T12:33:18.89139358Z",
        "message": "",
        "reason": "",
        "status": "False",
        "type": "OverlappingSelectors"
      }
    ]
  }
}
```

Device
```
{
  "apiVersion": "v1alpha1",
  "kind": "Device",
  "metadata": {
    "annotations": {
      "device-controller/renderedVersion": "2",
      "fleet-controller/templateVersion": "2024-08-06T12:33:18Z"
    },
    "creationTimestamp": "2024-08-06T12:33:39.288011Z",
    "generation": 2,
    "labels": {
      "alias": "device-a",
      "device": "device-a"
    },
    "name": "abc632d8633782719874da4052f957faa742fc7050026748bc79065c8819d1b",
    "owner": "Fleet/empty-fleet",
    "resourceVersion": "8"
  },
  "spec": {
    "config": []
  },
  "status": {
    "applications": {
      "data": {},
      "summary": {
        "status": "Unknown"
      }
    },
    "conditions": [
      {
        "lastTransitionTime": "2024-08-06T12:33:39.300770393Z",
        "message": "",
        "reason": "Valid",
        "status": "True",
        "type": "SpecValid"
      }
    ],
    "config": {
      "renderedVersion": ""
    },
    "integrity": {
      "summary": {
        "status": "Unknown"
      }
    },
    "lastSeen": "0001-01-01T00:00:00Z",
    "os": {
      "image": ""
    },
    "resources": {
      "cpu": "Unknown",
      "disk": "Unknown",
      "memory": "Unknown"
    },
    "summary": {
      "status": "Unknown"
    },
    "systemInfo": {
      "architecture": "",
      "bootID": "",
      "operatingSystem": ""
    },
    "updated": {
      "status": "Unknown"
    }
  }
}
```

Reply:
Nice catch!  Should be fixed now